### ISSUE TITLE
How to extract encoded features from an adversarial autoencoder (AAE)

### ISSUE THREAD

**Issue by lixiangchun** [wiseodd/generative-models#6]
Hello Wiseodd,

The source code of GAN and its variations provided here are very helpful for those who are interested in applying GAN in bioinformatics.

My question is how to extract encoded features from an AAE (e.g. aae_pytorch.py in the VAE folder). After reading the codes in aae_pytorch.py, I guess that the following codes can be used to extracted the encoded features of training set of MNIST data.

```python
X = Variable(torch.from_numpy(mnist.train.images))
z_fake = Q(X)
np.savetxt('encoded_feature_MNIST.txt', z_fake.data.numpy())
````

Am I correct?

Thanks,

Xiangchun

**Comment by wiseodd** [comment]
Hi Xiangchun. Yes, that's correct.

**Comment by lixiangchun** [comment]
Hi Wiseodd,

Thanks for your reply. Just another follow-up discussion.  When applying AAE to my data (consisted of 272 samples and each sample has 10386 number of features), I observed that the `D_loss` and `recon_loss` are consistently decreasing whereas the `G_loss` is increasing. 

Any idea on:
1. How to determine whether an AAE has been trained good enough
2. How to balance `D_loss`, `G_loss` and `recon_loss` simultaneously

Any comments and suggestions are appreciated.

Best,
Xiangchun

**Comment by wiseodd** [comment]
In AAE, reconstruction loss is well understood (it's just cross entropy or L2), and expected to decrease.

However, for GAN loss (`D_loss` and `G_loss`), there is no correlation between the quality of samples and the value of the loss. So, visual inspection is unfortunately still the best way to monitor the training.

To my knowledge, only recently there is a work that shows a meaningful loss function for GAN (WGAN paper <https://arxiv.org/abs/1701.07875>).

**Comment by lixiangchun** [comment]
Thanks for your suggestion.
